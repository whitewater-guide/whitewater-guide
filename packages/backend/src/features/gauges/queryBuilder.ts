import { buildListQuery, buildRootQuery, ListQueryBuilderOptions, QueryBuilderOptions } from '../../db';
import { Gauge } from '../../ww-commons';

// Not yet implemented
const customFieldMap = {
  lastMeasurement: () => 'script',
  status: () => 'script',
};

export const buildGaugeQuery = (options: Partial<QueryBuilderOptions<Gauge>>) =>
  buildRootQuery({
    info: options.info!,
    context: options.context!,
    table: 'gauges_view',
    customFieldMap,
    ...options,
  });

export const buildGaugesListQuery = (options: Partial<ListQueryBuilderOptions<Gauge>>) =>
  buildListQuery({
    info: options.info!,
    context: options.context!,
    table: 'gauges_view',
    customFieldMap,
    ...options,
  });
