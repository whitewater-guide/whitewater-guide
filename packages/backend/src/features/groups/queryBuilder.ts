import Knex from 'knex';
import db, { buildRootQuery, QueryBuilderOptions } from '../../db';
import { Group } from '../../ww-commons';

const connections = {
  regions: {
    getBuilder: () => require('../regions').buildRegionQuery,
    join: (table: string, query: Knex.QueryBuilder) => {
      const groupId = db(true).raw('??', ['groups_view.id']);
      return query
        .innerJoin('regions_groups', `${table}.id`, 'regions_groups.region_id')
        .where('regions_groups.group_id', '=', groupId);
    },
  },
};

export const buildGroupsListQuery = (options: Partial<QueryBuilderOptions<Group>>) =>
  buildRootQuery({
    info: options.info!,
    context: options.context!,
    table: 'groups_view',
    connections,
    orderBy: 'name',
    ...options,
  });
