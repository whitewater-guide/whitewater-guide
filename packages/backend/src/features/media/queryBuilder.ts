import { buildListQuery, buildRootQuery, ListQueryBuilderOptions, QueryBuilderOptions } from '@db';
import { Media } from '@ww-commons';
import Knex from 'knex';

const customFieldMap = {
  deleted: () => null,
};

export const buildMediaQuery = (options: Partial<QueryBuilderOptions<Media>>): Knex.QueryBuilder =>
  buildRootQuery({
    info: options.info!,
    context: options.context!,
    table: 'media_view',
    orderBy: 'weight',
    customFieldMap,
    ...options,
  });

export const buildMediaListQuery = (options: Partial<ListQueryBuilderOptions<Media>>) =>
  buildListQuery({
    info: options.info!,
    context: options.context!,
    table: 'media_view',
    orderBy: 'weight',
    customFieldMap,
    ...options,
  });
