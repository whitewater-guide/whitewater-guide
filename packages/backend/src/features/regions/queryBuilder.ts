import Knex from 'knex';
import db, { buildListQuery, buildRootQuery, ListQueryBuilderOptions, QueryBuilderOptions } from '../../db';
import { Region } from '../../ww-commons';

const customFieldMap = {
  editable: () => null,
  hasPremiumAccess: () => null,
  coverImage: () => 'cover_image',
};

const connections = {
  rivers: {
    getBuilder: () => require('../rivers').buildRiverQuery,
    foreignKey: 'region_id',
    join: (table: string, query: Knex.QueryBuilder) => {
      const regionId = db(true).raw('??', ['regions_view.id']);
      return query.where(`${table}.region_id`, '=', regionId);
    },
  },
  gauges: {
    getBuilder: () => require('../gauges').buildGaugeQuery,
    foreignKey: 'source_id',
    join: (table: string, query: Knex.QueryBuilder) => {
      const regionId = db(true).raw('??', ['regions_view.id']);
      return query
        .innerJoin('sources_regions', `${table}.source_id`, 'sources_regions.source_id')
        .where('sources_regions.region_id', '=', regionId);
    },
  },
  sections: {
    getBuilder: () => require('../sections').buildSectionQuery,
    foreignKey: 'river_id',
    join: (table: string, query: Knex.QueryBuilder) => {
      const regionId = db(true).raw('??', ['regions_view.id']);
      return query
        .innerJoin('rivers', `${table}.river_id`, 'rivers.id')
        .where('rivers.region_id', '=', regionId);
    },
  },
};

export const buildRegionQuery = (options: Partial<QueryBuilderOptions<Region>>): Knex.QueryBuilder =>
  buildRootQuery({
    context: options.context!,
    table: 'regions_view',
    customFieldMap,
    connections,
    ...options,
  });

export const buildRegionsListQuery = (options: Partial<ListQueryBuilderOptions<Region>>) =>
  buildListQuery({
    context: options.context!,
    table: 'regions_view',
    customFieldMap,
    connections,
    ...options,
  });
