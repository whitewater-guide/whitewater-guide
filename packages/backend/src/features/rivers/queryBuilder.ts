import * as Knex from 'knex';
import db, { buildListQuery, buildRootQuery, ListQueryBuilderOptions, QueryBuilderOptions } from '../../db';
import { River } from '../../ww-commons';

const oneToOnes = {
  region: {
    getBuilder: () => require('../regions').buildRegionQuery,
    join: (table: string, query: Knex.QueryBuilder) => {
      const regionId = db(true).raw('??', ['rivers_view.region_id']);
      return query.where(`${table}.id`, '=', regionId);
    },
  },
};

const connections = {
  sections: {
    getBuilder: () => require('../sections').buildSectionQuery,
    foreignKey: 'river_id',
    join: (table: string, query: Knex.QueryBuilder) => {
      const riverId = db(true).raw('??', ['rivers_view.id']);
      return query.where(`${table}.river_id`, '=', riverId);
    },
  },
};

export const buildRiverQuery = (options: Partial<QueryBuilderOptions<River>>): Knex.QueryBuilder =>
  buildRootQuery({
    info: options.info!,
    context: options.context!,
    table: 'rivers_view',
    connections,
    oneToOnes,
    ...options,
  });

export const buildRiversListQuery = (options: Partial<ListQueryBuilderOptions<River>>) =>
  buildListQuery({
    info: options.info!,
    context: options.context!,
    table: 'rivers_view',
    connections,
    oneToOnes,
    ...options,
  });
