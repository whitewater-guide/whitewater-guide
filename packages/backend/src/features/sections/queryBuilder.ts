import Knex, { QueryBuilder } from 'knex';
import db, { buildListQuery, buildRootQuery, ListQueryBuilderOptions, QueryBuilderOptions } from '../../db';
import { Section } from '../../ww-commons';

const connections = {
  media: {
    getBuilder: () => require('../media').buildMediaQuery,
    join: (table: string, query: Knex.QueryBuilder) => {
      const sectionId = db(true).raw('??', ['sections_view.id']);
      return query
        .innerJoin('sections_media', `${table}.id`, 'sections_media.media_id')
        .where('sections_media.section_id', '=', sectionId);
    },
  },
};

const oneToOnes = {
  river: {
    getBuilder: () => require('../rivers').buildRiverQuery,
    join: (table: string, query: Knex.QueryBuilder) => {
      const riverId = db(true).raw('??', ['sections_view.river_id']);
      return query.where(`${table}.id`, '=', riverId);
    },
  },
  gauge: {
    getBuilder: () => require('../gauges').buildGaugeQuery,
    join: (table: string, query: Knex.QueryBuilder) => {
      const gaugeId = db(true).raw('??', ['sections_view.gauge_id']);
      return query.where(`${table}.id`, '=', gaugeId);
    },
  },
  region: {
    getBuilder: () => require('../regions').buildRegionQuery,
    join: (table: string, query: Knex.QueryBuilder) => {
      const riverId = db(true).raw('??', ['sections_view.river_id']);
      return query
        .innerJoin('rivers', `${table}.id`, 'rivers.region_id')
        .where('rivers.id', '=', riverId);
    },
  },
};

export const buildSectionQuery = (options: Partial<QueryBuilderOptions<Section>>) =>
  buildRootQuery({
    context: options.context!,
    table: 'sections_view',
    oneToOnes,
    connections,
    ...options,
  });

export const buildSectionsListQuery = (options: Partial<ListQueryBuilderOptions<Section>>) => {
  const query = buildListQuery({
    context: options.context!,
    table: 'sections_view',
    oneToOnes,
    connections,
    ...options,
  });

  // Filter hidden regions
  if (!options.context!.user) {
    query.where('sections_view.hidden', false);
  } else if (!options.context!.user!.admin) {
    query.where((qb) => {
      qb
        .where('sections_view.hidden', false)
        .orWhereIn('sections_view.region_id', function(this: QueryBuilder) {
          this.select('regions_editors.region_id')
            .from('regions_editors')
            .where('regions_editors.user_id', options.context!.user!.id);
        });
    });
  }
  return query;
};
