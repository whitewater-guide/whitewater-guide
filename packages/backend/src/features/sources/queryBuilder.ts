import * as Knex from 'knex';
import db, { buildListQuery, buildRootQuery, ListQueryBuilderOptions, QueryBuilderOptions } from '../../db';
import { Source } from '../../ww-commons';

const customFieldMap = {
  status: () => 'script',
};

const connections = {
  regions: {
    getBuilder: () => require('../regions').buildRegionQuery,
    join: (table: string, query: Knex.QueryBuilder) => {
      const sourceId = db(true).raw('??', ['sources_view.id']);
      return query
        .innerJoin('sources_regions', `${table}.id`, 'sources_regions.region_id')
        .where('sources_regions.source_id', '=', sourceId);
    },
  },
  gauges: {
    getBuilder: () => require('../gauges').buildGaugeQuery,
    foreignKey: 'source_id',
    join: (table: string, query: Knex.QueryBuilder) => {
      const sourceId = db(true).raw('??', ['sources_view.id']);
      return query.where(`${table}.source_id`, '=', sourceId);
    },
  },
};

export const buildQuery = (options: Partial<QueryBuilderOptions<Source>>) =>
  buildRootQuery({
    context: options.context!,
    table: 'sources_view',
    connections,
    customFieldMap,
    ...options,
  });

export const buildSourcesListQuery = (options: Partial<ListQueryBuilderOptions<Source>>) =>
  buildListQuery({
    context: options.context!,
    table: 'sources_view',
    customFieldMap,
    connections,
    ...options,
  });
