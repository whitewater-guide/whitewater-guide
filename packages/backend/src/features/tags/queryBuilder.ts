import { buildRootQuery, QueryBuilderOptions } from '../../db';
import { Tag } from '../../ww-commons';

export const buildTagsListQuery = (options: Partial<QueryBuilderOptions<Tag>>) =>
  buildRootQuery({
    info: options.info!,
    context: options.context!,
    table: 'tags_view',
    orderBy: 'name',
    ...options,
  });
