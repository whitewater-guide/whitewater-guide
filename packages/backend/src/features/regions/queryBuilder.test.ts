import { GraphQLResolveInfo } from 'graphql';
import gqf from 'graphql-fields';
import db from '../../db';
import { adminContext } from '../../test/context';
import { buildRegionQuery, buildRegionsListQuery } from './queryBuilder';

const graphqlFields: jest.Mock<any> = gqf as any;
jest.mock('graphql-fields', () => jest.fn());

const primitives = {
  __typename: {},
  id: {},
  language: {},
  name: {},
  description: {},
  season: {},
  seasonNumeric: {},
  bounds: {},
  pois: {},
  hidden: {},
  createdAt: {},
  updatedAt: {},
};

const info: GraphQLResolveInfo = {} as any;

const options = { info, context: adminContext, knex: db(true) };

describe('details', () => {
  it('should build correct query without connections', () => {
    graphqlFields.mockReturnValueOnce(primitives);
    const query = buildRegionQuery(options);
    expect(query).toMatchSnapshot();
  });

  it('should build correct query with rivers connection', () => {
    const connections = {
      rivers: {
        nodes: {
          __typename: {},
          id: {},
          language: {},
          name: {},
        },
        count: {},
      },
    };
    graphqlFields.mockReturnValueOnce({ ...primitives, ...connections });
    const query = buildRegionQuery(options);
    expect(query).toMatchSnapshot();
  });

  it('should build correct query with gauges connection', () => {
    const connections = {
      gauges: {
        nodes: {
          __typename: {},
          id: {},
          language: {},
          name: {},
        },
        count: {},
      },
    };
    graphqlFields.mockReturnValueOnce({ ...primitives, ...connections });
    const query = buildRegionQuery(options);
    expect(query).toMatchSnapshot();
  });

  it('should build correct query with sections connection', () => {
    const connections = {
      sections: {
        nodes: {
          __typename: {},
          id: {},
          language: {},
          name: {},
        },
        count: {},
      },
    };
    graphqlFields.mockReturnValueOnce({ ...primitives, ...connections });
    const query = buildRegionQuery(options);
    expect(query).toMatchSnapshot();
  });
});

describe('list', () => {
  it('should build correct query without connections', () => {
    graphqlFields.mockReturnValueOnce({ nodes: primitives, count: {} });
    const query = buildRegionsListQuery(options);
    expect(query).toMatchSnapshot();
  });

  it('should build correct query with rivers connection', () => {
    const connections = {
      rivers: {
        count: {},
      },
    };
    graphqlFields.mockReturnValueOnce({ nodes: { ...primitives, ...connections }, count: {} });
    const query = buildRegionsListQuery(options);
    expect(query).toMatchSnapshot();
  });

  it('should build correct query with gauges connection', () => {
    const connections = {
      gauges: {
        count: {},
      },
    };
    graphqlFields.mockReturnValueOnce({ nodes: { ...primitives, ...connections }, count: {} });
    const query = buildRegionsListQuery(options);
    expect(query).toMatchSnapshot();
  });

  it('should build correct query with sections connection', () => {
    const connections = {
      sections: {
        count: {},
      },
    };
    graphqlFields.mockReturnValueOnce({ nodes: { ...primitives, ...connections }, count: {} });
    const query = buildRegionsListQuery(options);
    expect(query).toMatchSnapshot();
  });
});
