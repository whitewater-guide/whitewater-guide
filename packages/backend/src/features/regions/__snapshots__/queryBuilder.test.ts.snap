// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`details should build correct query with gauges connection 1`] = `
with gauges_connection as (
    select
        gauges_view.id,
        gauges_view.name,
        gauges_view.source_id
    from
        gauges_view
    where
        gauges_view.language = 'en'
    order by
        gauges_view.name asc,
        gauges_view.created_at asc,
        gauges_view.id asc
)
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at, (with gauges_connection_internal as (
            select
                gauges_connection.*,
                COALESCE(count(gauges_connection.*)
                    OVER (),
                    0)
            from
                gauges_connection
            inner join sources_regions on gauges_connection.source_id = sources_regions.source_id
            where
                sources_regions.region_id = regions_view.id
)
        select
            json_build_object('count', (
                    select
                        gauges_connection_internal.count from gauges_connection_internal
                    limit 1),
                'nodes', (
                    select
                        COALESCE(json_agg (gauges_connection_internal.*),
                            '[]'::json)
                    from
                        gauges_connection_internal))) as gauges
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;

exports[`details should build correct query with rivers connection 1`] = `
with rivers_connection as (
    select
        rivers_view.id,
        rivers_view.name,
        rivers_view.region_id
    from
        rivers_view
    where
        rivers_view.language = 'en'
    order by
        rivers_view.name asc,
        rivers_view.created_at asc,
        rivers_view.id asc
)
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at, (with rivers_connection_internal as (
            select
                rivers_connection.*,
                COALESCE(count(rivers_connection.*)
                    OVER (),
                    0)
            from
                rivers_connection
            where
                rivers_connection.region_id = regions_view.id
)
        select
            json_build_object('count', (
                    select
                        rivers_connection_internal.count from rivers_connection_internal
                    limit 1),
                'nodes', (
                    select
                        COALESCE(json_agg (rivers_connection_internal.*),
                            '[]'::json)
                    from
                        rivers_connection_internal))) as rivers
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;

exports[`details should build correct query with sections connection 1`] = `
with sections_connection as (
    select
        sections_view.id,
        sections_view.name,
        sections_view.river_id
    from
        sections_view
    where
        sections_view.language = 'en'
    order by
        sections_view.name asc,
        sections_view.created_at asc,
        sections_view.id asc
)
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at, (with sections_connection_internal as (
            select
                sections_connection.*,
                COALESCE(count(sections_connection.*)
                    OVER (),
                    0)
            from
                sections_connection
            inner join rivers on sections_connection.river_id = rivers.id
            where
                rivers.region_id = regions_view.id
)
        select
            json_build_object('count', (
                    select
                        sections_connection_internal.count from sections_connection_internal
                    limit 1),
                'nodes', (
                    select
                        COALESCE(json_agg (sections_connection_internal.*),
                            '[]'::json)
                    from
                        sections_connection_internal))) as sections
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;

exports[`details should build correct query without connections 1`] = `
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;

exports[`list should build correct query with gauges connection 1`] = `
with gauges_connection as (
    select
        gauges_view.id,
        gauges_view.source_id
    from
        gauges_view
    where
        gauges_view.language = 'en'
    order by
        gauges_view.name asc,
        gauges_view.created_at asc,
        gauges_view.id asc
)
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at, (with gauges_connection_internal as (
            select
                gauges_connection.*,
                COALESCE(count(gauges_connection.*)
                    OVER (),
                    0)
            from
                gauges_connection
            inner join sources_regions on gauges_connection.source_id = sources_regions.source_id
            where
                sources_regions.region_id = regions_view.id
)
        select
            json_build_object('count', (
                    select
                        gauges_connection_internal.count from gauges_connection_internal
                    limit 1))) as gauges,
count(regions_view.*)
OVER ()
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;

exports[`list should build correct query with rivers connection 1`] = `
with rivers_connection as (
    select
        rivers_view.id,
        rivers_view.region_id
    from
        rivers_view
    where
        rivers_view.language = 'en'
    order by
        rivers_view.name asc,
        rivers_view.created_at asc,
        rivers_view.id asc
)
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at, (with rivers_connection_internal as (
            select
                rivers_connection.*,
                COALESCE(count(rivers_connection.*)
                    OVER (),
                    0)
            from
                rivers_connection
            where
                rivers_connection.region_id = regions_view.id
)
        select
            json_build_object('count', (
                    select
                        rivers_connection_internal.count from rivers_connection_internal
                    limit 1))) as rivers,
count(regions_view.*)
OVER ()
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;

exports[`list should build correct query with sections connection 1`] = `
with sections_connection as (
    select
        sections_view.id,
        sections_view.river_id
    from
        sections_view
    where
        sections_view.language = 'en'
    order by
        sections_view.name asc,
        sections_view.created_at asc,
        sections_view.id asc
)
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at, (with sections_connection_internal as (
            select
                sections_connection.*,
                COALESCE(count(sections_connection.*)
                    OVER (),
                    0)
            from
                sections_connection
            inner join rivers on sections_connection.river_id = rivers.id
            where
                rivers.region_id = regions_view.id
)
        select
            json_build_object('count', (
                    select
                        sections_connection_internal.count from sections_connection_internal
                    limit 1))) as sections,
count(regions_view.*)
OVER ()
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;

exports[`list should build correct query without connections 1`] = `
select
    regions_view.id,
    regions_view.name,
    regions_view.description,
    regions_view.season,
    regions_view.season_numeric,
    regions_view.bounds,
    regions_view.pois,
    regions_view.hidden,
    regions_view.created_at,
    regions_view.updated_at,
    count(regions_view.*)
    OVER ()
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc,
    regions_view.created_at asc,
    regions_view.id asc

`;
